documentation/projects/proposals/data_normalization/20240227-implementation_plan_catalog_data_cleaning.md

```diff
@@ -0,0 +1,198 @@
+| 2024-02-13 04:41:22 | 22155 |    9035451    |        8809204        |   0    |
+
+To have some numbers of the problem we are delaing with, the previous table
+shows the number of records cleaned by field for last runs at the moment of
+writing this IP, except for tags, which we don't have accurate registries since
+file saving was disabled.
+
+### Make and run a batched update DAG for one-time cleanup
+
+A batched catalog cleaner DAG (or potentially a `batched_update_from_file`)
+should take the files of the previous step to perform an batched update on the
+catalog's image table, while handling deadlocking and timeout concerns, similar
+to the [batched_update][batched_update]. This table is constantly in use by
+other DAGs, such as those from API providers or the data refresh process, and
+ideally can't be singly blocked by any DAG.
+
+[batched_update]: ./../../../catalog/reference/DAGs.md#batched_update
+
+A [proof of concept PR](https://github.com/WordPress/openverse/pull/3601)
+consisted of uploading each file to temporary `UNLOGGED` DB tables (which
+provides huge gains in writing performance while their disadventages are not
+relevant since they won't be permanent), and including a `row_id` serial number
+used later to query it in batches. Adding an index in this last column after
+filling up the table could improve the query performance. An adaptation will be
+needed to handle the column type of tags (`jsonb`).
+
+### Run an image data refresh to confirm cleaning time is reduced
+
+Finally, after the previous steps are done, running a data refresh will confirm
+there are no more updates applied at ingestion. If time isn't significantly
+reduced then it will be necessary to check what was missing in the previous
+steps.
+
+If confirmed the time is reduced to zero, optionally the cleaning steps can be
+removed, or leave them in case we want to perform a similar cleaning effort
+later.
+
+## Dependencies
+
+### Infrastructure
+
+No changes needed. The Ingestion Server already has the credentials required to
+[connect with AWS](https://github.com/WordPress/openverse/blob/a930ee0f1f116bac77cf56d1fb0923989613df6d/ingestion_server/ingestion_server/indexer_worker.py#L23-L28).
+
+<!--
+### Tools & packages
+
+ Describe any tools or packages which this work might be dependent on. If multiple options are available, try to list as many as are reasonable with your own recommendation. -->
+
+### Other projects or work
+
+Once the steps have been completed and proved the method works we could make
+additional similar corrections following the same procedure. Some potentially
+related issues are:
+
+- [Some images have duplicate incorrectly decoded unicode tags #1303](https://github.com/WordPress/openverse/issues/1303)
+- [Provider scripts may include html tags in record titles #1441](https://github.com/WordPress/openverse/issues/1441)
+- [Fix Wikimedia image titles #1728](https://github.com/WordPress/openverse/issues/1728)
```

Comment:
Thanks for noting these!

Reply:
Another thing we could add here is `license_url` check: save all of the items with their computed `license_url` values if this URL is not set in metadata. I suspect that there might be none such items, in which case we can simplify this part of API:

https://github.com/WordPress/openverse/blob/3c5cdd069bc0a9e2aa1f979bb99a7ee1402f8560/api/api/models/media.py#L86-L92

Here's a related issue: https://github.com/WordPress/openverse/issues/552
Note that there are also probably some items with incorrectly set license pairs with metadata set to `null` that I didn't finish fixing in https://github.com/WordPress/openverse-catalog/pull/1045

Reply:
One more ingestion server step we can do is remove the extraction of the extension from the image URL: https://github.com/WordPress/openverse/issues/1560

These two things are not part of the cleanup process per se, so we might want to add them to the cleanup once, specifically for saving the fixed information, and remove these steps (and the code in the ES models` as soon as we confirm that all data has been cleaned.

Reply:
@obulat That's a good observation on the `license_url`! I'll run a query to see if there are still items without it.

I'll add the file type/extension issue to the list ➕

Reply:
Thank you for investigating, @krysal .

For the item you mention, it doesn't have `raw_license_url`, but it does have `license_url`, so it's fine. However, if you do `SELECT * FROM image WHERE meta_data->>'license_url' is null`, you get a lot of results.

I think we could run DAG from [Add a DAG for backfilling license_url when meta_data is null](https://github.com/WordPress/openverse-catalog/pull/1005) for items where `meta_data.license_url` is null. And only resort to the process from this PR if that DAG takes too long.